pkg/sqlcache/informer/listoption_indexer.go

```diff
@@ -0,0 +1,896 @@
+	}
+	o, ok := a.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object type, expected unstructured.Unstructured: %v", a)
+	}
+
+	var obj interface{}
+	var found bool
+	obj = o.Object
+	for i, subField := range subFields {
+		switch t := obj.(type) {
+		case map[string]interface{}:
+			subField = strings.TrimSuffix(strings.TrimPrefix(subField, "["), "]")
+			obj, found, err = unstructured.NestedFieldNoCopy(t, subField)
+			if err != nil {
+				return nil, err
+			}
+			if !found {
+				// particularly with labels/annotation indexes, it is totally possible that some objects won't have these,
+				// so either we this is not an error state or it could be an error state with a type that callers can check for
+				return nil, nil
+			}
+		case []interface{}:
+			if strings.HasPrefix(subField, "[") && strings.HasSuffix(subField, "]") {
+				key, err := strconv.Atoi(strings.TrimSuffix(strings.TrimPrefix(subField, "["), "]"))
+				if err != nil {
+					return nil, fmt.Errorf("[listoption indexer] failed to convert subfield [%s] to int in listoption index: %w", subField, err)
+				}
+				if key >= len(t) {
+					return nil, fmt.Errorf("[listoption indexer] given index is too large for slice of len %d", len(t))
+				}
+				obj = fmt.Sprintf("%v", t[key])
+			} else if i == len(subFields)-1 {
+				result := make([]string, len(t))
+				for index, v := range t {
+					itemVal, ok := v.(map[string]interface{})
+					if !ok {
+						return nil, fmt.Errorf(failedToGetFromSliceFmt, subField, err)
+					}
+					itemStr, ok := itemVal[subField].(string)
+					if !ok {
+						return nil, fmt.Errorf(failedToGetFromSliceFmt, subField, err)
+					}
+					result[index] = itemStr
+				}
+				return result, nil
+			}
+		default:
+			return nil, fmt.Errorf("[listoption indexer] failed to parse subfields: %v", subFields)
+		}
+	}
+	return obj, nil
+}
+
+func extractSubFields(fields string) ([]string, error) {
+	var subFields []string
+
+	for {
+		idx := strings.IndexAny(fields, "[.")
+		if idx < 0 {
+			if len(fields) > 0 {
+				subFields = append(subFields, fields)
+			}
+			return subFields, nil
+		}
+		switch fields[idx] {
+		case '[':
+			// At this point we have something like:
+			// foo[bar]
+			//    ^-- idx
+			// so we must make a subfield for foo
+			subField := fields[:idx]
+			subFields = append(subFields, subField)
+			fields = fields[idx:]
+
+			// And then we search for the closing parenthese and make
+			// a subfield for [bar] (parens included)
+			parentIdx := strings.IndexByte(fields, ']')
+			if parentIdx < 0 {
+				return nil, fmt.Errorf("missing closing parentheses for field %s", fields)
+			}
+			subField = fields[:parentIdx+1]
+			subFields = append(subFields, subField)
+			fields = fields[parentIdx+1:]
+			if len(fields) > 0 && fields[0] == '.' {
+				fields = fields[1:]
+			}
+		case '.':
+			subField := fields[:idx]
+			subFields = append(subFields, subField)
+			fields = fields[idx+1:]
+		}
+	}
+}
```

Comment:
These lines do not seem to exist in lasso main:


https://github.com/rancher/lasso/blob/9e2b68739ccc04fe6c080fc0489391970d77dd29/pkg/cache/sql/informer/listoption_indexer.go#L828-L834

(at the time of writing main points at 9e2b6 in lasso)


I could not track down which branch these lines come from.